plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -981,23 +986,30 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
                         tableMetadatas.add(TableColumnsMetadata.forTable(tableName, columns));
                     });
 
-                    for (SchemaTableName tableName : remainingTables) {
-                        try {
-                            Table icebergTable = catalog.loadTable(session, tableName);
-                            List<ColumnMetadata> columns = getColumnMetadatas(icebergTable.schema(), typeManager);
-                            tableMetadatas.add(TableColumnsMetadata.forTable(tableName, columns));
-                        }
-                        catch (TableNotFoundException e) {
-                            // Table disappeared during listing operation
-                        }
-                        catch (UnknownTableTypeException e) {
-                            // Skip unsupported table type in case that the table redirects are not enabled
-                        }
-                        catch (RuntimeException e) {
-                            // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
-                            log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
-                        }
-                    }
+                    List<CompletableFuture<Void>> futures = remainingTables.stream()
+                            .map(tableName -> CompletableFuture.runAsync(() -> {
+                                try {
+                                    Table icebergTable = catalog.loadTable(session, tableName);
+                                    List<ColumnMetadata> columns = getColumnMetadatas(icebergTable.schema(), typeManager);
+                                    synchronized (tableMetadatas) {
+                                        tableMetadatas.add(TableColumnsMetadata.forTable(tableName, columns));
+                                    }
+                                }
+                                catch (TableNotFoundException e) {
+                                    // Table disappeared during listing operation
+                                }
+                                catch (UnknownTableTypeException e) {
+                                    // Skip unsupported table type in case that the table redirects are not enabled
+                                }
+                                catch (RuntimeException e) {
+                                    // Table can be being removed and this may cause all sorts of exceptions. Log, because we're catching broadly.
+                                    log.warn(e, "Failed to access metadata of table %s during streaming table columns for %s", tableName, prefix);
+                                }
+                            }, metadataFetchingExecutor))
+                            .toList();
+
+                    CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
```

Comment:
How will this behave if the query is cancelled? Will pending tasks also be cancelled? I’d suggest using `ExecutorUtil#processWithAdditionalThreads` instead, as it handles such scenarios.

Reply:
I have updated the code to use`processWithAdditionalThreads `